shared/domains/chelonia/persistent-actions.js

```diff
@@ -0,0 +1,200 @@
+  errorInvocation?: SbpInvocation,
+  // Maximum number of tries, default: Infinity.
+  maxAttempts: number,
+  // How many seconds to wait between retries.
+  retrySeconds: number,
+  skipCondition?: SbpInvocation,
+  // SBP selector to call on success with the received value.
+  successInvocationSelector?: string,
+  totalFailureInvocation?: SbpInvocation
+}
+
+type PersistentActionStatus = {
+  cancelled: boolean,
+  failedAttemptsSoFar: number,
+  lastError: string,
+  nextRetry: string,
+  pending: boolean
+}
+
+const defaultOptions: PersistentActionOptions = {
+  maxAttempts: Number.POSITIVE_INFINITY,
+  retrySeconds: 30
+}
+const tag = '[chelonia.persistentActions]'
+
+class PersistentAction {
+  invocation: SbpInvocation
+  options: PersistentActionOptions
+  status: PersistentActionStatus
+  timer: Object
+
+  constructor (invocation: SbpInvocation, options: PersistentActionOptions = {}) {
+    this.invocation = invocation
+    this.options = { ...defaultOptions, ...options }
+    this.status = {
+      cancelled: false,
+      failedAttemptsSoFar: 0,
+      lastError: '',
+      nextRetry: '',
+      pending: false
+    }
+  }
+
+  // Do not call if the action is pending or cancelled!
+  async attempt (): Promise<void> {
+    if (await this.trySBP(this.options.skipCondition)) {
+      this.cancel()
+      return
+    }
+    try {
+      this.status.pending = true
+      const result = await sbp(...this.invocation)
+      this.handleSuccess(result)
+    } catch (error) {
+      this.handleError(error)
+    }
+  }
+
+  cancel (): void {
+    this.timer && clearTimeout(this.timer)
+    this.status.cancelled = true
+    this.status.nextRetry = ''
+  }
+
+  async handleError (error: Error): Promise<void> {
+    const { options, status } = this
+    // Update relevant status fields before calling any optional selector.
+    status.failedAttemptsSoFar++
+    status.lastError = error.message
+    const anyAttemptLeft = options.maxAttempts > status.failedAttemptsSoFar
+    status.nextRetry = anyAttemptLeft && !status.cancelled
+      ? new Date(Date.now() + options.retrySeconds * 1e3).toISOString()
+      : ''
+    // Perform any optional SBP invocation.
+    await this.trySBP(options.errorInvocation)
+    !anyAttemptLeft && await this.trySBP(options.totalFailureInvocation)
+    // Schedule a retry if appropriate.
+    if (status.nextRetry) {
+      // Note: there should be no older active timeout to clear.
+      this.timer = setTimeout(() => this.attempt(), this.options.retrySeconds * 1e3)
+    }
+    status.pending = false
+  }
+
+  async handleSuccess (result: any): Promise<void> {
+    const { status } = this
+    status.lastError = ''
+    status.nextRetry = ''
+    status.pending = false
+    this.options.successInvocationSelector &&
+      await this.trySBP([this.options.successInvocationSelector, result])
+  }
+
+  trySBP (invocation: SbpInvocation | void): any {
+    try {
+      return invocation ? sbp(...invocation) : undefined
```

Comment:
there is no `await` on the call to `sbp` so if an exception occurs on a selector that's async it won't be caught

Reply:
Yes, but you can still do that if you `await` here.

By not calling `await` here, it defeats the purpose of this function. You might as well remove `trySBP` (personally I would remove it and directly wrap calls that might fail in `try/catch` at their call site).

The `try/catch` here won't get triggered if there's an exception thrown here by `sbp` for a selector that is async. Instead there will be an unhandled promise error shown in the console. try/catch only works with `await` when dealing with async functions.

Reply:
> I would remove it and directly wrap calls that might fail in `try/catch` at their call site).
We can do that, but that's going to require four `try/catch` blocks instead of one